common/src/main/java/io/opentelemetry/sdk/trace/FieldBackedSpanValueStorageProvider.java

```diff
@@ -0,0 +1,104 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.opentelemetry.sdk.trace;
+
+import co.elastic.otel.common.SpanValueStorage;
+import co.elastic.otel.common.SpanValueStorageProvider;
+import io.opentelemetry.api.trace.Span;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * This class enables {@link co.elastic.otel.common.SpanValue}s to be stored directly as fields on
+ * spans. The field ($elasticSpanValues) is injected at packaging time via the shadow plugin to our
+ * agent distro.
+ *
+ * <p>This class needs to live in the same package as the OpenTelemetry SdkSpan, otherwise it is not
+ * possible to create an {@link AtomicReferenceFieldUpdater} for safely initializing the field.
+ */
+@SuppressWarnings("unchecked")
+public class FieldBackedSpanValueStorageProvider implements SpanValueStorageProvider {
+
+  private static final String FIELD_NAME = "$elasticSpanValues";
+
+  private static final Logger logger = Logger.getLogger(SpanValueStorage.class.getName());
+
+  public static final FieldBackedSpanValueStorageProvider INSTANCE;
+
+  private static final MethodHandle spanFieldGetter;
+  private static final AtomicReferenceFieldUpdater<SdkSpan, Object> spanFieldSetter;
+
+  static {
+    FieldBackedSpanValueStorageProvider resultInstance = null;
+    MethodHandle getter = null;
+    AtomicReferenceFieldUpdater<SdkSpan, Object> setter = null;
+    try {
+      Field storageField = SdkSpan.class.getDeclaredField(FIELD_NAME);
+      if (storageField.getType() != Object.class) {
+        throw new IllegalStateException("Unexpected field type: " + storageField.getType());
+      }
+
+      getter = MethodHandles.lookup().unreflectGetter(storageField);
+      setter = AtomicReferenceFieldUpdater.newUpdater(SdkSpan.class, Object.class, FIELD_NAME);
+
+      logger.log(Level.FINE, "Using field-backed storage for SpanValues", FIELD_NAME);
+      resultInstance = new FieldBackedSpanValueStorageProvider();
+
+    } catch (NoSuchFieldException e) {
+      logger.log(
+          Level.FINE,
+          "Using map-backed storage for SpanValues because Field '{0}' does not exist on SdkSpan",
+          FIELD_NAME);
```

Comment:
This is intentionally on debug level because this is the expected case when running our extensions outside of the agent with a vanilla SDK. If anything else goes wrong, we throw an error in the catch clause below.

At the same time we should catch problems with our code modification with the integration test I added in this PR.

Reply:
That's a very good point, I forgot about usage of this with just an extension, so the SDK code modification would not be applied and then the fallback solution is needed to have something working (albeit with few compromises on performance).